pkg/lang/ir/v1/system.go

```diff
@@ -46,6 +47,92 @@ func (g generalGraph) compileUbuntuAPT(root llb.State) llb.State {
 	return root
 }
 
+// copyAPTSignature returns the state and the path in string format
+// The returned string represents the location of the apt-source signature
+// A successful run of copyAPTSignature should return path of /etc/apt/keyrings/*.asc
+func (g generalGraph) copyAPTSignature(root llb.State, name string, url string) (llb.State, string) {
+
+	var fileName = fmt.Sprintf("%s.asc", name)
+	var fileFolder = "/etc/apt/keyrings"
+
+	// path stores the location of the signature file
+	// The value of path should be /etc/apt/keyrings/*.asc
+	var path = filepath.Join(fileFolder, fileName)
+
+	base := llb.Image(builderImage)
+	builder := base.
+		Run(llb.Shlex(fmt.Sprintf("sh -c \"curl %s >> %s\"", url, fileName)),
+			llb.WithCustomName("[internal] downloading apt-source signature in base image")).Root()
+
+	aptSign := root.
+		File(llb.Mkdir(fileFolder, 0755, llb.WithParents(true)),
+			llb.WithCustomName("[internal] setting target apt-source signature folder")).
+		File(llb.Copy(builder, fileName, path),
+			llb.WithCustomName("[internal] copy signature from builder"))
+
+	return aptSign, path
+}
+
+// configRSrc returns the state and the content in DEB822 format for third-party apt-source
+// The returned string contains all configuration for adding third-party into apt source list
+// A successful run of configRSrc should return the content strictly follow the DEB822 format
+func (g generalGraph) configRSrc(root llb.State, aptConfig ir.APTConfig, sign string) (llb.State, string) {
+
+	var enabled = fmt.Sprintf("Enabled: %s\n", aptConfig.Enabled)
+	var types = fmt.Sprintf("Types: %s\n", aptConfig.Types)
+	var uris = fmt.Sprintf("URIs: %s\n", aptConfig.URIs)
+	var suites = fmt.Sprintf("Suites: %s\n", aptConfig.Suites)
+	var components = fmt.Sprintf("Components: %s\n", aptConfig.Components)
+	var architecture = fmt.Sprintf("Architectures: %s\n", aptConfig.Arch)
+
+	aptSign, signPath := g.copyAPTSignature(root, aptConfig.Name, sign)
+	var signature = fmt.Sprintf("Signed-By: %s\n", signPath)
+
+	var content strings.Builder
+	content.WriteString(enabled)
+	content.WriteString(types)
+	content.WriteString(uris)
+	content.WriteString(suites)
+	content.WriteString(components)
+	content.WriteString(signature)
+	content.WriteString(architecture)
+
+	return aptSign, content.String()
+}
+
+// compileRLang returns the llb.State only after compoiling the environment for installing R language
+// A successful run of compileRLang should set up the official R apt repo into /etc/apt/sources.list.d/*.sources
+func (g generalGraph) compileRLang(root llb.State) llb.State {
+
+	// sign stores the third-party URI for downloading the signature of the corresponding repo
+	var sign = "https://cloud.r-project.org/bin/linux/ubuntu/marutter_pubkey.asc"
```

Comment:
This could be a const.

Reply:
Fixed

Reply:
The const can be placed at the top of this file. Refer to other files in the ir package.

Reply:
Fixec